cmd/repo-init/api_test.go

```diff
@@ -0,0 +1,184 @@
+				initConfig{
+					Org:                   "org",
+					Repo:                  "repo",
+					Branch:                "branch",
+					CanonicalGoRepository: "sometimes.com",
+					GoVersion:             "1",
+				},
+			},
+			validationType: All,
+			expected: &validationResponse{
+				Valid: false,
+				ValidationErrors: []validationError{
+					{
+						Key:     "generic",
+						Message: "invalid configuration: you must define at least one test or image build in 'tests' or 'images'",
+					},
+				},
+			},
+		},
+		{
+			name: "Validate operator substitution - valid",
+			data: SubstitutionValidationRequest{
+				ConfigValidationRequest: ConfigValidationRequest{
+					initConfig{
+						Org:                   "org",
+						Repo:                  "repo",
+						Branch:                "branch",
+						CanonicalGoRepository: "sometimes.com",
+						GoVersion:             "1",
+						OperatorBundle: &operatorBundle{
+							DockerfilePath: "Dockerfile.bundle",
+							Name:           "ci-index",
+						},
+					},
+				},
+				Substitution: api.PullSpecSubstitution{
+					PullSpec: "test.io/pullspec:4.1",
+					With:     "image1",
+				},
+			},
+			validationType: OperatorSubstitution,
+			expected: &validationResponse{
+				Valid: true,
+			},
+		},
+		{
+			name: "Validate operator substitution - invalid",
+			data: SubstitutionValidationRequest{
+				ConfigValidationRequest: ConfigValidationRequest{
+					initConfig{
+						Org:                   "org",
+						Repo:                  "repo",
+						Branch:                "branch",
+						CanonicalGoRepository: "sometimes.com",
+						GoVersion:             "1",
+						OperatorBundle: &operatorBundle{
+							DockerfilePath: "Dockerfile.bundle",
+							Name:           "ci-index",
+						},
+					},
+				},
+				Substitution: api.PullSpecSubstitution{
+					PullSpec: "test.io/pullspec:4.1",
+					With:     "thisaintright:image1",
+				},
+			},
+			validationType: OperatorSubstitution,
+			expected: &validationResponse{
+				Valid: false,
+				ValidationErrors: []validationError{
+					{
+						Field:   "operator_substitution",
+						Message: "with: could not resolve 'thisaintright:image1' to an image involved in the config",
+					},
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			data, _ := json.Marshal(tc.data)
+			marshalled, _ := json.Marshal(ValidationRequest{
+				ValidationType: tc.validationType,
+				Data:           data,
+			})
+			body := bytes.NewBuffer(marshalled)
+
+			r, err := http.NewRequest(http.MethodGet, "wordup.com", body)
+			if err != nil {
+				t.Fatalf("could not make request: %v", err)
+			}
+
+			writer := &fakeWriter{}
+
+			validateConfig(writer, r)
+
+			actual := &validationResponse{}
+			_ = json.Unmarshal(writer.body, actual)
+			if diff := cmp.Diff(tc.expected, actual); diff != "" {
+				t.Errorf("%s: got invalid response: %v", tc.name, diff)
+			}
+		})
+	}
+
+}
+
+type fakeWriter struct {
```

Comment:
httptesting package may have what you need

Reply:
Ah, cool, let me take a look. Was looking through existing tests to see if I could find something re-usable but didn't have much luck.